p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, _ := multihash.Encode(localMh, multihash.SHA2_256)
+	localFpMultibase, _ := multibase.Encode(multibase.Base64url, localMhBuf)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := udpmux.NewUDPMux(socket, func(ufrag string, addr net.Addr) {
+		// Push to the candidateChan asynchronously to avoid blocking the mux goroutine
+		// on candidates being processed. This can cause new connections to fail at high
+		// throughput but will allow packets for existing connections to be processed.
+		select {
+		case candidateChan <- candidateAddr{ufrag: ufrag, raddr: addr.(*net.UDPAddr)}:
+		default:
+			log.Debug("candidate chan full, dropping incoming candidate")
+		}
+	})
+
+	l := &listener{
+		mux:                       mux,
+		transport:                 transport,
+		config:                    config,
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		localMultiaddr:            laddr,
+		ctx:                       ctx,
+		cancel:                    cancel,
+		localAddr:                 socket.LocalAddr(),
+		acceptQueue:               make(chan tpt.CapableConn, transport.maxInFlightConnections),
+		inFlightConnections:       0,
+		maxInFlightConnections:    transport.maxInFlightConnections,
+	}
+
+	l.wg.Add(1)
+	go l.handleIncomingCandidates(candidateChan)
+	return l, err
+}
+
+func (l *listener) handleIncomingCandidates(candidateChan chan candidateAddr) {
+	defer l.wg.Done()
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+		case addr := <-candidateChan:
+			if atomic.LoadUint32(&l.inFlightConnections) >= l.maxInFlightConnections {
+				// TODO: should we send an error STUN response here? It seems like Pion and browsers will retry
+				// STUN binding requests even when an error response is received.
+				// Refer: https://github.com/pion/ice/blob/master/agent.go#L1045-L1131
+				log.Warnf("server is busy, rejecting incoming connection from: %s", addr.raddr)
+				continue
+			}
+			atomic.AddUint32(&l.inFlightConnections, 1)
+			go func() {
+				ctx, cancel := context.WithTimeout(l.ctx, candidateSetupTimeout)
+				defer cancel()
+				conn, err := l.handleCandidate(ctx, addr)
+				if err != nil {
+					log.Debugf("could not accept connection: %s: %v", addr.ufrag, err)
+					atomic.AddUint32(&l.inFlightConnections, ^uint32(0))
+					return
+				}
+				select {
+				case l.acceptQueue <- conn:
+				default:
+					log.Warnf("could not push connection")
+					atomic.AddUint32(&l.inFlightConnections, ^uint32(0))
+					conn.Close()
+				}
+			}()
+		}
+	}
+}
+
+func (l *listener) Accept() (tpt.CapableConn, error) {
+	select {
+	case <-l.ctx.Done():
+		return nil, os.ErrClosed
+	case conn := <-l.acceptQueue:
+		atomic.AddUint32(&l.inFlightConnections, ^uint32(0))
+		return conn, nil
+	}
+}
+
+func (l *listener) Close() error {
+	select {
+	case <-l.ctx.Done():
+		return nil
+	default:
+	}
+	l.cancel()
```

Comment:
why the preference (as I see it in multiple places) to have this "default" logic below the select, rather than as part of the "default" statement? o.O I mean it probably nets out to the same, but I wonder why you prefer this, given you anyway already have the body ready to be used?

Reply:
In these cases I see use of the select block similar to polling the channel to see if it has data or if it is closed, and doing something else otherwise.

Reply:
Sure I get that. What I do not get is why you write the code as-is,
instead of the following:

```go
func (l *listener) Close() error {
 	select {
 	case <-l.ctx.Done():
 	default:
		l.cancel()
		l.wg.Wait()
 	}
 	return nil
}
```

Reply:
Isn't that the exact same end result?